src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php

```diff
@@ -137,28 +146,45 @@ public function dispatch($eventName, Event $event = null)
         $this->preDispatch($eventName, $event);
 
         $e = $this->stopwatch->start($eventName, 'section');
+        $previousRequestHash = $this->currentRequestHash;
 
-        $this->dispatcher->dispatch($eventName, $event);
+        try {
+            $this->dispatcher->dispatch($eventName, $event);
+            $this->currentRequestHash = $previousRequestHash;
 
-        if ($e->isStarted()) {
-            $e->stop();
-        }
+            if ($e->isStarted()) {
+                $e->stop();
+            }
 
-        $this->postDispatch($eventName, $event);
-        $this->postProcess($eventName);
+            $this->postDispatch($eventName, $event);
+        } finally {
+            $this->currentRequestHash = $previousRequestHash;
+
+            if ($e->isStarted()) {
+                $e->stop();
+            }
+
+            $this->postProcess($eventName);
+        }
 
         return $event;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getCalledListeners()
+    public function getCalledListeners(/* Request $request = null */)
```

Comment:
All added arguments should be declared explicitly in the docblock (see eg src/Symfony/Bridge/Monolog/Logger.php)